keps/sig-instrumentation/draft-20190404-kubernetes-control-plane-metrics-stability.md

```diff
@@ -0,0 +1,248 @@
+  - sig-instrumentation
+  - sig-api-machinery
+  - sig-node
+reviewers:
+  - TBD
+approvers:
+  - TBD
+editor: TBD
+creation-date: 2019-04-04
+see-also:
+  - 0031-kubernetes-metrics-overhaul
+status: provisional
+---
+
+# Kubernetes Control-Plane Metrics Stability
+
+## Table of Contents
+
+* [Table of Contents](#table-of-contents)
+* [Summary](#summary)
+* [Motivation](#motivation)
+    * [Goals](#goals)
+    * [Non-Goals](#non-goals)
+* [Background](#background)
+* [Proposal](#proposal)
+    * [Metric Definition Phase](#metric-definition-phase)
+    * [Metric Instantiation Phase](#metric-instantiation-phase)
+    * [Metric Enrollment Phase](#metric-enrollment-phase)
+* [Design Details](#design-details)
+  * [Test Plan](#test-plan)
+  * [Graduation Criteria](#graduation-criteria)
+* [Drawbacks](#drawbacks)
+* [Alternative](#alternatives)
+* [Unresolved Questions](#unresolved-questions)
+* [Implementation History](#implementation-history)
+* [Reference](#reference)
+
+## Summary
+
+Currently metrics emitted in the kubernetes control-plane do not offer any stability guarantees. This Kubernetes Enhancement Proposal (KEP) proposes a strategy and framework for programmatically expressing how stable a metric is. This KEP also defines the specific guarantees made for each enumerated level of stability. Since this document will likely evolve with ongoing discussion around metric stability, it will be updated accordingly.
+
+## Motivation
+
+Metrics stability has been an ongoing community concern. Oftentimes, cluster monitoring infrastructure assumes the stability of at least some control-plane metrics; thus, it would be prudent to offer some sort of guarantees around control-plane metrics, treating it more properly as an API. Since the [metrics overhaul](https://github.com/kubernetes/enhancements/blob/master/keps/sig-instrumentation/0031-kubernetes-metrics-overhaul.md) is nearing completion, there should be less reason to introduce breaking changes to metrics, making it an opportune time to introduce metric stability rules. Specifically, this KEP intends to address metric stability from an ingestion point of view.
+
+Guarantees around metrics have been [proposed previously](https://docs.google.com/document/d/1_CdNWIjPBqVDMvu82aJICQsSCbh2BR-y9a8uXjQm4TI/edit#) and there are [ongoing community discussions](https://groups.google.com/forum/#!topic/kubernetes-sig-instrumentation/XbElxDtww0Y) around this issue. Some suggested solutions include:
+
+1. Having a ‘stable’ metrics endpoint, i.e. ‘/metrics/v1’
+2. Leaving metrics as is and documenting the ones which have a stability guarantee
+
+This KEP suggests another alternative but is very much in line with the spirit of the other proposed solutions.
+
+### Goals
+
+ * Describe the various stability guarantees for control-plane metrics.
+ * Define a uniform mechanism for expressing metric stability.
+
+### Non-Goals
+
+* We are __*not*__ defining which specific metrics are actually stable.
```

Comment:
explicitly mention *here* that these are for control plane metrics, and not resource metrics, which (potentially) need a separate discussion

cc @dashpole

Reply:
Sure thing.

Reply:
I agree that resource metrics are in a different boat from most metrics we are considering here.  Ill be out for a while, so i'll leave my thoughts here.

I think using what is proposed here with versioned endpoints is actually a good model.  Since this proposal essentially marks each metric as alpha/ga, I don't think there is any use in having, for example, a v1alpha1 endpoint.  Each individual metric already comes with a contractual guarantee for stability, which presumably also means the endpoint itself has a contractual guarantee.  If we can't remove a metric foo for 2 releases, we can't remove the endpoint it is on for at least that long.

However, there could be a case where we want to replace all metrics on a given endpoint.  One example might be switching the metrics to a different format (OpenMetrics?).  In that case, it would be useful to have a` /v1` endpoint, which can exist alongside a `/v2` endpoint, and clearly separate the two groups of metrics which have clear differentiating properties.

However, I think it is also perfectly acceptable to just omit changes to the kubelet's resource metrics endpoint from this proposal.